lib/puppet/provider/iis_virtual_directory/webadministration.rb

```diff
@@ -18,47 +19,71 @@ def exists?
-      fail "physicalpath is a required paramter to create a iis virtual directory"
-    end
+    verify_physicalpath
 
-    if @resource[:physicalpath] and ! File.exists?(@resource[:physicalpath])
-      fail "physicalpath doesn't exist: #{@resource[:physicalpath]}"
-    end
+    # New-WebVirtualDirectory fails when PhysicalPath is a UNC path that is not available:
+    #   "Parameter 'PhysicalPath' should point to existing path."
 
-    cmd = []
-    cmd << "New-WebVirtualDirectory -Name \"#{@resource[:name]}\" "
-    cmd << "-Site \"#{@resource[:sitename]}\" " if @resource[:sitename]
-    cmd << "-Application \"#{@resource[:application]}\" " if @resource[:application]
-    cmd << "-PhysicalPath \"#{@resource[:physicalpath]}\" " if @resource[:physicalpath]
-    cmd << "-ErrorAction Stop"
-    cmd = cmd.join
+    if is_drive_path(@resource[:physicalpath])
+      cmd = []
+      cmd << "New-WebVirtualDirectory -Name \"#{@resource[:name]}\" "
+      cmd << "-Site \"#{@resource[:sitename]}\" " if @resource[:sitename]
+      cmd << "-Application \"#{@resource[:application]}\" " if @resource[:application]
+      cmd << "-PhysicalPath \"#{@resource[:physicalpath]}\" "
+      cmd << "-ErrorAction Stop"
+      cmd = cmd.join
 
-    result   = self.class.run(cmd)
-    Puppet.err "Error creating virtual directory: #{result[:errormessage]}" unless result[:exitcode] == 0
-    @resource[:ensure] = :present
+      result   = self.class.run(cmd)
+      Puppet.err "Error creating virtual directory: #{result[:errormessage]}" unless result[:exitcode] == 0
+      @resource[:ensure] = :present
+    end
+    if is_unc_path(@resource[:physicalpath])
+      cmd = []
+      cmd << "New-Item -type VirtualDirectory \"IIS:\\Sites\\#{@resource[:sitename]}\\#{@resource[:name]}\" "
+      cmd << "-Application \"#{@resource[:application]}\" " if @resource[:application]
+      cmd << "-PhysicalPath \"#{@resource[:physicalpath]}\" "
+      cmd = cmd.join
+
+      result = self.class.run(cmd)
+      Puppet.err "Error creating virtual directory: #{result[:errormessage]}" unless result[:exitcode] == 0
+      @resource[:ensure] = :present
+    end
   end
 
   def update
+    # WORKAROUND: update() is called after destroy() by flush() in ../iis_powershell.rb
+    # New-Item -Force recreates the VirtualDirectory unless we return early.
+    return if (@resource[:ensure] == :absent)
+
     Puppet.debug "Updating #{@resource[:name]}"
 
-    if @resource[:physicalpath] and ! File.exists?(@resource[:physicalpath])
-      fail "physicalpath doesn't exist: #{@resource[:physicalpath]}"
+    verify_physicalpath
+
+    # Set-ItemProperty fails when PhysicalPath is a UNC path that is not available:
+    #   "Parameter 'PhysicalPath' should point to existing path."
+
+    if is_drive_path(@resource[:physicalpath])
+      cmd = "Set-ItemProperty 'IIS:\\Sites\\#{@resource[:sitename]}\\#{@resource[:name]}' -Name 'PhysicalPath' -Value '#{@resource[:physicalpath]}'"
```

Comment:
Again, this would be easier to read if we only replaced the cmdlet instead of the whole block, ew wouldn't have to duplicate the cmd execution part

Reply:
True.

Should the PowerShell Set-* commands in update() also have "-ErrorAction Stop"?

Reply:
Yes

Reply:
Adding ErrorAction stop in update() reveals there are no failures when setting `site` (or `application`):

```
Debug: /Stage[main]/Main/Node[pe-201722-agent-win2012]/Iis_virtual_directory[local]: Nothing to manage: no ensure and the resource doesn't exist                   
Notice: /Stage[main]/Main/Node[pe-201722-agent-win2012]/Iis_virtual_directory[test]/physicalpath: physicalpath changed '\\SERVER\shareXXX' to '\\SERVER\share'        
Debug: Updating test                                                                                                                                                
Debug: COMMAND: Set-ItemProperty -Path 'IIS:\Sites\Default Web Site\test' -Name 'physicalPath' -Value '\\SERVER\share' -ErrorAction Stop                            
Set-ItemProperty -Path 'IIS:\Sites\Default Web Site\test' -Name 'site' -Value 'Default Web Site' -ErrorAction Stop                                                  
Debug: Waited 50 milliseconds...                                                                                                                                   
Debug: Waited 50 milliseconds...                                                                                                                                   
Debug: Wait object signaled                                                                                                                                        
Debug: Waited 0.140081 total seconds.                                                                                                                              
Debug: 2017-11-17 01:35:54 +0000 PIPE> <ReturnResult><Property Name="exitcode">1</Property><Property Name="errormessage">UHJvcGVydHkgc2l0ZSBpcyBub3QgZm91bmQgb24gIF
xcUEUtMjAxNzIyLUFHRU5UXFNpdGVzXERlZmF1bHQgV2ViIFNpdGVcdW5jLg0KUGFyYW1ldGVyIG5hbWU6IHByb3BOYW1lDQo=</Property><Property Name="stderr"></Property><Property Name="std
out"></Property></ReturnResult>                                                                                                                                    
                                                                                                                                                                   
Debug: ERRMSG: Property site is not found on  \\PE-201722-AGENT\Sites\Default Web Site\test.                                                                        
Parameter name: propName                                                                                                                                           
                                                                                                                                                                   
Error: Error updating virtual directory: Property site is not found on  \\PE-201722-AGENT\Sites\Default Web Site\test.                                              
Parameter name: propName                                                                                                                                           
```

Get-ItemProperty confirms those attributes do not exist to set: 

```
PS C:\> Get-ItemProperty 'IIS:\Sites\Default Web Site\test' | Select *                                                                                              
                                                                                                                     
path              : /test                                                                                                                                           
physicalPath      : \\SERVER\share                                                                                                                                 
userName          :                                                                                                                                                
password          :                                                                                                                                                
logonMethod       : ClearText                                                                                                                                      
allowSubDirConfig : True                                                                                                                                           
ItemXPath         : /system.applicationHost/sites/site[@name='Default Web Site' and @id='1']/application[@path='/']/virtualDirectory[@path='/test']                 
Name              : unc                                                                                                                                            
PSPath            : WebAdministration::\\PE-201722-AGENT\Sites\Default Web Site\test                                                                                
PSParentPath      : WebAdministration::\\PE-201722-AGENT\Sites\Default Web Site                                                                                    
PSChildName       : test                                                                                                                                            
PSDrive           : IIS                                                                                                                                            
PSProvider        : WebAdministration                                                                                                                              
Attributes        : {path, physicalPath, userName, password...}                                                                                                    
ChildElements     : {}                                                                                                                                             
ElementTagName    : virtualDirectory                                                                                                                               
Methods           :                                                                                                                                                
Schema            : Microsoft.IIs.PowerShell.Framework.ConfigurationElementSchema                                                                                  
```